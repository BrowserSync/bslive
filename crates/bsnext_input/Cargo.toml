[package]
name = "bsnext_input"
version = "0.2.3"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
bsnext_resp = { path = "../bsnext_resp" }
bsnext_tracing = { path = "../bsnext_tracing" }
bsnext_guards = { path = "../bsnext_guards" }

miette = { workspace = true }

random_word = { workspace = true }
matchit = { workspace = true }
mime_guess = { workspace = true }
serde = { workspace = true }
clap = { workspace = true }
anyhow = { workspace = true }
serde_yaml = { workspace = true }
serde_json = { workspace = true }
toml = { workspace = true }
thiserror = { workspace = true }

[dev-dependencies]
insta = { workspace = true }
