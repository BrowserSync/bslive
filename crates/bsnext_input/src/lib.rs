use crate::target::TargetKind;

use std::fmt::{Display, Formatter};
use std::fs;
use std::fs::read_to_string;
use std::net::AddrParseError;
use std::path::{Path, PathBuf};

use crate::md::MarkdownError;
use crate::yml::YamlError;

#[cfg(test)]
pub mod input_test;
pub mod md;
pub mod paths;
pub mod route;
pub mod route_manifest;
pub mod server_config;
pub mod target;
#[cfg(test)]
pub mod watch_opt_test;
pub mod yml;

#[derive(Debug, Default, serde::Deserialize, serde::Serialize)]
pub struct Input {
    pub servers: Vec<server_config::ServerConfig>,
}

impl Input {
    pub fn from_input_path<P: AsRef<Path>>(path: P) -> Result<Self, InputError> {
        match path.as_ref().extension().and_then(|x| x.to_str()) {
            None => Err(InputError::MissingExtension(path.as_ref().to_owned())),
            Some("yml") | Some("yaml") => Input::from_yaml_path(path),
            Some("md") | Some("markdown") => Input::from_md_path(path),
            Some(other) => Err(InputError::UnsupportedExtension(other.to_string())),
        }
    }
    fn from_yaml_path<P: AsRef<Path>>(path: P) -> Result<Self, InputError> {
        let str = read_to_string(&path)?;
        let output = serde_yaml::from_str::<Self>(str.as_str()).map_err(move |e| {
            if let Some(location) = e.location() {
                YamlError::ParseErrorWithLocation {
                    serde_error: e,
                    input: str,
                    path: path.as_ref().to_string_lossy().to_string(),
                    line: location.line(),
                    column: location.column(),
                }
            } else {
                YamlError::ParseError {
                    serde_error: e,
                    input: str,
                    path: path.as_ref().to_string_lossy().to_string(),
                }
            }
        })?;
        // todo: don't allow duplicates?.
        Ok(output)
    }
    fn from_md_path<P: AsRef<Path>>(path: P) -> Result<Self, InputError> {
        let str = read_to_string(path)?;
        let input = md::md_to_input(&str)?;
        // todo: don't allow duplicates.
        Ok(input)
    }
}

#[derive(Debug, thiserror::Error)]
pub enum InputError {
    #[error("no suitable inputs could be found")]
    MissingInputs,
    #[error("could not read input, error: {0}")]
    InvalidInput(String),
    #[error("io error")]
    Io(#[from] std::io::Error),
    #[error("Could not find the input file: {0}")]
    NotFound(PathBuf),
    #[error("Paths without extensions are not supported: {0}")]
    MissingExtension(PathBuf),
    #[error("Unsupported extension: {0}")]
    UnsupportedExtension(String),
    #[error("InputWriteError prevented startup {0}")]
    InputWriteError(#[from] InputWriteError),
    #[error("Input path error prevented startup {0}")]
    PathError(#[from] PathError),
    #[error("Input port error prevented startup {0}")]
    PortError(#[from] PortError),
    #[error("Input directory error prevented startup {0}")]
    DirError(#[from] DirError),
    #[error("Markdown error: {0}")]
    MarkdownError(#[from] MarkdownError),
    #[error("{0}")]
    YamlError(#[from] YamlError),
}

#[derive(Debug, thiserror::Error)]
pub enum WatchError {
    #[error("don't add `.` before the extension.")]
    InvalidExtensionFilter,
    #[error("empty")]
    EmptyExtensionFilter,
}

#[derive(Debug, thiserror::Error)]
pub enum InputWriteError {
    #[error("couldn't write input to {path}")]
    FailedWrite { path: PathBuf },
}

#[derive(Debug, thiserror::Error, serde::Serialize, serde::Deserialize)]
pub enum PathError {
    #[error("path(s) not found \n{paths}")]
    MissingPaths { paths: PathDefs },
}

#[derive(Debug, thiserror::Error)]
#[error(transparent)]
pub enum PortError {
    #[error("could not use that port: {port} {err}")]
    InvalidPort { port: u16, err: AddrParseError },
}

#[derive(Debug, thiserror::Error)]
#[error(transparent)]
pub enum DirError {
    #[error("could not create that dir: {path}")]
    CannotCreate { path: PathBuf },
    #[error("could not change the process CWD to: {path}")]
    CannotMove { path: PathBuf },
}

#[derive(Debug, thiserror::Error, serde::Serialize, serde::Deserialize)]
pub struct PathDefs(Vec<PathDefinition>);

#[derive(Debug, thiserror::Error, serde::Serialize, serde::Deserialize)]
struct PathDefinition {
    pub input: String,
    pub cwd: PathBuf,
    pub absolute: PathBuf,
}

impl Display for PathDefs {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        for pd in self.0.iter() {
            writeln!(f, "  cwd:   {}", pd.cwd.display())?;
            writeln!(f, "  input: {}", pd.input)?;
        }
        Ok(())
    }
}
impl Display for PathDefinition {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("PathDefinition").finish()
    }
}

pub fn rand_word() -> String {
    random_word::gen(random_word::Lang::En).to_string()
}

pub fn fs_write_input(
    cwd: &Path,
    input: &Input,
    target_kind: TargetKind,
) -> Result<PathBuf, InputWriteError> {
    let string = match target_kind {
        TargetKind::Yaml => serde_yaml::to_string(&input).expect("create yaml?"),
        TargetKind::Toml => toml::to_string_pretty(&input).expect("create toml?"),
        TargetKind::Md => md::input_to_str(input),
    };
    let name = match target_kind {
        TargetKind::Yaml => "input.yml",
        TargetKind::Toml => "input.toml",
        TargetKind::Md => "input.md",
    };
    let next_path = cwd.join(name);
    fs::write(&next_path, string)
        .map(|()| next_path.clone())
        .map_err(|_e| InputWriteError::FailedWrite { path: next_path })
}
