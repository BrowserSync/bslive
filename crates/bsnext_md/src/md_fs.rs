use crate::md_to_input;
use bsnext_input::{Input, InputCreation, InputError};
use std::fs::read_to_string;
use std::path::Path;

pub struct MdFs;

impl InputCreation for MdFs {
    fn from_input_path<P: AsRef<Path>>(path: P) -> Result<Input, Box<InputError>> {
        let str = read_to_string(path).map_err(|e| Box::new(e.into()))?;
        let input =
            md_to_input(&str).map_err(|e| Box::new(InputError::MarkdownError(e.to_string())))?;
        Ok(input)
    }
    fn from_input_str<P: AsRef<str>>(content: P) -> Result<Input, Box<InputError>> {
        let input = md_to_input(content.as_ref())
            .map_err(|e| Box::new(InputError::MarkdownError(e.to_string())))?;
        Ok(input)
    }
}
