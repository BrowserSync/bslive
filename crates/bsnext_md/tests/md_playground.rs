use bsnext_input::InputCreation;
use bsnext_md::md_fs::MdFs;

#[test]
fn test_md_playground() -> anyhow::Result<()> {
    // let input = include_str!("../../examples/markdown/markdown.md");
    let input = r#"

```html playground

<div class="main">
    Hello world!
</div>
```

```css
@import url("/reset.css");

:root {
    border: 50px solid pink;
    height: 100vh;
    overflow: hidden;
}
```

```js
console.log("hello world")
```

        "#;
    let config = MdFs::from_input_str(&input, &Default::default()).expect("unwrap");
    let first_server = config.servers.get(0).unwrap();
    let routes = first_server
        .playground
        .as_ref()
        .map(|x| x.as_routes().unwrap())
        .unwrap();
    insta::assert_debug_snapshot!(routes);
    Ok(())
}
