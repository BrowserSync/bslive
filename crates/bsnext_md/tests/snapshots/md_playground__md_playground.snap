---
source: crates/bsnext_md/tests/md_playground.rs
expression: routes
---
[
    Route {
        path: PathDef {
            inner: "/",
        },
        kind: Raw(
            Html {
                html: "\n<div class=\"main\">\n    Hello world!\n</div>",
            },
        ),
        opts: Opts {
            cors: None,
            delay: None,
            watch: Bool(
                true,
            ),
            inject: Items(
                [
                    InjectionItem {
                        inner: Addition(
                            InjectAddition {
                                addition_position: Prepend(
                                    "\n<!doctype html>\n  <html lang=\"en\">\n  <head>\n      <meta charset=\"UTF-8\">\n      <meta name=\"viewport\" content=\"width=device-width, initial-scale=1\">\n      <title>Browsersync Live - Playground</title>\n      <link rel=\"stylesheet\" href=\"__bslive_playground.css\">\n  </head>\n  <body>\n",
                                ),
                            },
                        ),
                        only: Some(
                            Item(
                                Str(
                                    RootPath {
                                        inner: "/",
                                    },
                                ),
                            ),
                        ),
                    },
                    InjectionItem {
                        inner: Addition(
                            InjectAddition {
                                addition_position: Append(
                                    "\n  <script src=\"__bslive_playground.js\" type=\"module\"></script>\n  </body>\n</html>\n",
                                ),
                            },
                        ),
                        only: Some(
                            Item(
                                Str(
                                    RootPath {
                                        inner: "/",
                                    },
                                ),
                            ),
                        ),
                    },
                    InjectionItem {
                        inner: BsLive(
                            BuiltinStringDef {
                                name: Connector,
                            },
                        ),
                        only: Some(
                            Item(
                                Str(
                                    RootPath {
                                        inner: "/",
                                    },
                                ),
                            ),
                        ),
                    },
                ],
            ),
            headers: None,
            cache: Prevent,
            compression: None,
        },
        fallback: Some(
            FallbackRoute {
                kind: Raw(
                    Html {
                        html: "This is a BSLIVE playground",
                    },
                ),
                opts: Opts {
                    cors: None,
                    delay: None,
                    watch: Bool(
                        true,
                    ),
                    inject: Bool(
                        true,
                    ),
                    headers: None,
                    cache: Prevent,
                    compression: None,
                },
            },
        ),
        when: Some(
            WhenOne(
                ExactUri {
                    exact_uri: true,
                },
            ),
        ),
        when_body: None,
    },
    Route {
        path: PathDef {
            inner: "/__bslive_playground.js",
        },
        kind: Raw(
            Raw {
                raw: "console.log(\"hello world\")",
            },
        ),
        opts: Opts {
            cors: None,
            delay: None,
            watch: Bool(
                true,
            ),
            inject: Bool(
                true,
            ),
            headers: None,
            cache: Prevent,
            compression: None,
        },
        fallback: None,
        when: None,
        when_body: None,
    },
    Route {
        path: PathDef {
            inner: "/__bslive_playground.css",
        },
        kind: Raw(
            Raw {
                raw: "@import url(\"/reset.css\");\n\n:root {\n    border: 50px solid pink;\n    height: 100vh;\n    overflow: hidden;\n}",
            },
        ),
        opts: Opts {
            cors: None,
            delay: None,
            watch: Bool(
                true,
            ),
            inject: Bool(
                true,
            ),
            headers: None,
            cache: Prevent,
            compression: None,
        },
        fallback: None,
        when: None,
        when_body: None,
    },
    Route {
        path: PathDef {
            inner: "/reset.css",
        },
        kind: Raw(
            Raw {
                raw: "/*\n  1. Use a more-intuitive box-sizing model.\n*/\n*,\n*::before,\n*::after {\n    box-sizing: border-box;\n}\n\n/*\n  2. Remove default margin\n*/\n* {\n    margin: 0;\n}\n\n/*\n  Typographic tweaks!\n  3. Add accessible line-height\n  4. Improve text rendering\n*/\nbody {\n    line-height: 1.5;\n    -webkit-font-smoothing: antialiased;\n}\n\n/*\n  5. Improve media defaults\n*/\nimg,\npicture,\nvideo,\ncanvas,\nsvg {\n    display: block;\n    max-width: 100%;\n}\n\n/*\n  6. Remove built-in form typography styles\n*/\ninput,\nbutton,\ntextarea,\nselect {\n    font: inherit;\n}\n\n/*\n  7. Avoid text overflows\n*/\np,\nh1,\nh2,\nh3,\nh4,\nh5,\nh6 {\n    overflow-wrap: break-word;\n}\n\n/*\n  8. Create a root stacking context\n*/\n#root,\n#__next {\n    isolation: isolate;\n}\n",
            },
        ),
        opts: Opts {
            cors: None,
            delay: None,
            watch: Bool(
                true,
            ),
            inject: Bool(
                true,
            ),
            headers: None,
            cache: Prevent,
            compression: None,
        },
        fallback: None,
        when: None,
        when_body: None,
    },
]
